### PR TITLE
Epoch syncing - syncing only epoch blocks.

### PR DESCRIPTION
**Pull request by Frozen**
## Issue
Optimize beacon chain sync by only syncing the epoch blocks. 
https://github.com/harmony-one/harmony/issues/4052

## Test

https://gist.github.com/Frozen/fd0ffb735e94f7424416d5421c9341e1

## Operational Checklist

1. **Does this PR introduce backward-incompatible changes to the on-disk data structure and/or the over-the-wire protocol?**. (If no, skip to question 8.)

    **NO**

8. **Does this PR introduce backward-incompatible changes *NOT* related to on-disk data structure and/or over-the-wire protocol?** (If no, continue to question 11.)

    **NO**

### CODE REVIEW

File: api/service/legacysync/epoch_syncing.go [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+	}
+	// TODO(k.potapov) validate blocks
+
+	for _, block := range decoded {
+		_, err := bc.StoreShardStateBytes(block.Header().Epoch(), block.Header().ShardState())
+		if err != nil {
+			return err
+		}
+
+		err = bc.WriteBlockWithoutState(block, common.Big1)
```
Comment (JackyWYX) [inline api/service/legacysync/epoch_syncing.go:232]: Have you tested locally or on mainnet? Is there any error returned from this line?

Comment (JackyWYX) [reply api/service/legacysync/epoch_syncing.go:232]: There is a lot of messages written to leveldb other than the state. Furthermore, it seems there are not enough check on the block message.

Comment (Frozen) [reply api/service/legacysync/epoch_syncing.go:232]: Yes, it works locally, i'm going to test it on testnet. There are no checks on block inside WriteBlockWithoutState, i'll add signature validation, other impossible without state.

Comment (JackyWYX) [reply api/service/legacysync/epoch_syncing.go:232]: Please take a look at Finalize function at `internal/chain/engine.go:276`. You will at least need the state shard for the signature check.

File: api/service/legacysync/epoch_syncing.go [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+	}
+
+	for _, block := range decoded {
+		sig, bitmap, err := chain.ParseCommitSigAndBitmap(block.GetCurrentCommitSig())
+		if err != nil {
+			return errors.Wrap(err, "parse commitSigAndBitmap")
+		}
+
+		// Signature validation.
+		err = bc.Engine().VerifyHeaderSignature(bc, block.Header(), sig, bitmap)
```
Comment (Frozen) [inline api/service/legacysync/epoch_syncing.go:232]: I've added signature validation, but i'm not sure is it correct and enough? Maybe i should add some extra validation?

File: api/service/legacysync/epoch_syncing.go (hunk 1 above)
Comment (Frozen) [reply api/service/legacysync/epoch_syncing.go:242]: Yea, i checked that function, but it seems my changes does not affect it. Shard state functions called only for beacon chain, and my code works only for non-beacon.

File: api/service/legacysync/epoch_syncing.go (hunk 2 above)
Comment (rlan35) [reply api/service/legacysync/epoch_syncing.go:232]: Signature verification is enough and we can't validate other data since we don't have the consecutive blocks and state.

**Comment by JackyWYX** [comment]
For best engineering practice, I would personally suggest to write a `EpochChain` which maintains a header of epoch blocks and validators data. It can be a component within the `HeaderChain` to avoid doing nasty stuff outside the core module.

File: api/service/legacysync/epoch_syncing.go [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+			utils.Logger().Error().Err(err).
+				Msgf("[SYNC] ProcessStateSync failed (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d)",
+					isBeacon, bc.ShardID(), maxHeight, height)
+			return 2
+		}
+	}
+}
+
+// getMaxPeerHeight gets the maximum blockchain heights from peers
+func (ss *EpochSync) getMaxPeerHeight(isBeacon bool) uint64 {
```
Comment (rlan35) [inline api/service/legacysync/epoch_syncing.go:161]: Please refactor the code so no duplicated code like this exists.

File: api/service/legacysync/epoch_syncing.go [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+// If the last result is expired, ask the remote DNS nodes for latest height and return the result.
+func (ss *EpochSync) GetSyncStatus() SyncCheckResult {
+	return ss.syncStatus.Get(func() SyncCheckResult {
+		return ss.isInSync(false)
+	})
+}
+
+// isInSync query the remote DNS node for the latest height to check what is the current
+// sync status
+func (ss *EpochSync) isInSync(doubleCheck bool) SyncCheckResult {
```
Comment (rlan35) [inline api/service/legacysync/epoch_syncing.go:50]: Whether epoch blocks are in sync should depends on checking the latest epoch block right? Seems this part of the code is just a copy and paste of the logic that's checking the latest block of the chain?

File: api/service/legacysync/epoch_syncing.go [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+		curEpoch := block.Epoch()
+		for i := 0; i < int(SyncLoopBatchSize); i++ {
+			epochLastBlock := shard.Schedule.EpochLastBlock(curEpoch.Uint64())
+			if epochLastBlock > maxHeight {
+				break
+			}
+			if height != epochLastBlock {
+				heights = append(heights, epochLastBlock)
+			}
+			curEpoch = curEpoch.Add(curEpoch, common.Big1)
```
Comment (rlan35) [inline api/service/legacysync/epoch_syncing.go:142]: what is curEpoch used for?

File: api/service/legacysync/epoch_syncing.go (hunk 1 above)
Comment (rlan35) [reply api/service/legacysync/epoch_syncing.go:251]: You need to call WriteShardStateBytes() or something related to write the shardState to DB. Take a look at WriteBlockWithState and see how it works, especially the part about offChain data. Then you will know why WriteShardStateBytes() is needed. Because otherwise, the shard state can be read from the DB.

Comment (Frozen) [reply api/service/legacysync/epoch_syncing.go:251]: I use StoreShardStateBytes above, that function is a wrapper for `WriteShardStateBytes` and calls it internally. So, yes, i store shardState to DB. `WriteBlockWithoutState` i need for height in order to understand where i am right now and what to do next. Without `WriteBlockWithoutState` it starts to panic in different places, it's better not to skip that.

File: api/service/legacysync/epoch_syncing.go (hunk 4 above)
Comment (Frozen) [reply api/service/legacysync/epoch_syncing.go:49]: Right, now it depends on last epoch.

File: api/service/legacysync/epoch_syncing.go (hunk 3 above)
Comment (Frozen) [reply api/service/legacysync/epoch_syncing.go]: fixed

File: api/service/legacysync/epoch_syncing.go [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+		if otherHeight1 < lastHeight {
+			heightDiff = 0 //
+		}
+		onTop := epochLastBlock > otherHeight1
+		utils.Logger().Info().
+			Uint64("OtherHeight", otherHeight1).
+			Uint64("lastHeight", lastHeight).
+			Msg("[SYNC] Checking sync status")
+		return SyncCheckResult{
+			IsInSync:    !onTop,
```
Comment (rlan35) [inline api/service/legacysync/epoch_syncing.go:69]: onTop means our height is bigger than other height, so it means in sync right? So here it should be:

IsInSync: onTop.

File: api/service/legacysync/epoch_syncing.go [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+	"github.com/harmony-one/harmony/internal/chain"
+	"github.com/harmony-one/harmony/internal/utils"
+	"github.com/harmony-one/harmony/node/worker"
+	"github.com/harmony-one/harmony/p2p"
+	"github.com/harmony-one/harmony/shard"
+	"github.com/pkg/errors"
+)
+
+type EpochSync struct {
+	blockChain         *core.BlockChain
```
Comment (rlan35) [inline api/service/legacysync/epoch_syncing.go:23]: Since we are talking about epoch block sync for beacon chain, this blockChain should be beaconChain right?

File: api/service/legacysync/epoch_syncing.go (hunk 4 above)
Comment (rlan35) [reply api/service/legacysync/epoch_syncing.go:49]: I think for checking whether the epoch blocks are latest in sync with the beacon chain, the check logic should be:

1. Get the latest epoch block I have: myEpoch.
2. Get maxPeerHeight: otherHeight
3. using shard schedule to calculate what epoch is the "otherHeight" belong and get: otherEpoch.
4. Compare otherEpoch vs myEpoch to decide whether I am in sync or not.

Comment (Frozen) [reply api/service/legacysync/epoch_syncing.go:49]: OnTop means i have the highest block that possible. In epoch context it means that i have the highest epoch. I can calculate last block for this epoch, and if it doesn't exists yet means i'm idling. Otherwise i'm synchronizing. I do the same above, but in another words, without converting into epoch, but in heights.

File: api/service/legacysync/epoch_syncing.go (hunk 6 above)
Comment (Frozen) [reply api/service/legacysync/epoch_syncing.go:69]: onTop means that our possible height for last block is higher than the highest known block from network.

Comment (Frozen) [reply api/service/legacysync/epoch_syncing.go]: Fixed

File: api/service/legacysync/epoch_syncing.go (hunk 4 above)
Comment (Frozen) [reply api/service/legacysync/epoch_syncing.go:48]: i've realized that "in sync" means already synchronized but not "in synchronizing process", exactly opposite meaning. Fixed.

File: api/service/legacysync/epoch_syncing.go [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+}
+
+// isInSync query the remote DNS node for the latest height to check what is the current
+// sync status
+func (ss *EpochSync) isInSync(doubleCheck bool) SyncCheckResult {
+	if ss.syncConfig == nil {
+		return SyncCheckResult{} // If syncConfig is not instantiated, return not in sync
+	}
+	otherHeight1 := getMaxPeerHeight(ss.syncConfig)
+	curBlock := ss.beaconChain.CurrentBlock()
```
Comment (rlan35) [inline api/service/legacysync/epoch_syncing.go:53]: which block does this return? With epoch sync'ing, we don't have the full beacon chain, right? we only have all the epoch blocks of the beacon chain.

File: api/service/legacysync/epoch_syncing.go [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+	}
+
+	for _, block := range decoded {
+		sig, bitmap, err := chain.ParseCommitSigAndBitmap(block.GetCurrentCommitSig())
+		if err != nil {
+			return errors.Wrap(err, "parse commitSigAndBitmap")
+		}
+
+		// Signature validation.
+		err = bc.Engine().VerifyHeaderSignature(bc, block.Header(), sig, bitmap)
```
Comment (rlan35) [inline api/service/legacysync/epoch_syncing.go:195]: Actually, this verification won't work, because the commit/sig stored in the block N is for the last block N-1, to get the commit/sig for block N, you need to find them in the block N+1.

Comment (rlan35) [reply api/service/legacysync/epoch_syncing.go:195]: Have you tested this part of the code? I bet the verification of the sig should fail.

File: api/service/legacysync/epoch_syncing.go [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// getConsensusHashes gets all hashes needed to download.
+func (ss *EpochSync) getConsensusHashes(startHash []byte, size uint32) error {
```
Comment (rlan35) [inline api/service/legacysync/epoch_syncing.go:222]: this and CreateSyncConfig method are all copy/paste, please use inheritance to refactor them so it's reused.

File: api/service/legacysync/helpers.go [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+import (
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common/math"
+	"github.com/harmony-one/harmony/internal/utils"
+)
+
+// getMaxPeerHeight gets the maximum blockchain heights from peers
+func getMaxPeerHeight(syncConfig *SyncConfig) uint64 {
```
Comment (rlan35) [inline api/service/legacysync/helpers.go:11]: same here, duplicate code shouldn't exist.

File: core/blockchain.go [hunk 12]
```diff
@@ -334,15 +360,18 @@ func (bc *BlockChain) loadLastState() error {
...
-	if _, err := state.New(currentBlock.Root(), bc.stateCache); err != nil {
-		// Dangling block without a state associated, init from scratch
-		utils.Logger().Warn().
-			Str("number", currentBlock.Number().String()).
-			Str("hash", currentBlock.Hash().Hex()).
-			Msg("Head state missing, repairing chain")
-		if err := bc.repair(&currentBlock); err != nil {
-			return err
+
+	if !bc.options.SkipInitialStateValidation {
```
Comment (rlan35) [inline core/blockchain.go:364]: Why is this needed? please add comment to explain.

File: core/blockchain.go [hunk 13]
```diff
@@ -1912,6 +1945,22 @@ func (bc *BlockChain) WriteShardStateBytes(db rawdb.DatabaseWriter,
 	return decodeShardState, nil
 }
 
+func (bc *BlockChain) StoreShardStateBytes(epoch *big.Int, shardState []byte) (*shard.State, error) {
```
Comment (rlan35) [inline core/blockchain.go:1948]: can you reuse the existing WriteShardStateBytes() instead ?

**Comment by rlan35** [comment]
Is this code working for mainnet data? Please test by running a shard 1 explorer node.

File: api/service/legacysync/epoch_syncing.go (hunk 8 above)
Comment (Frozen) [reply api/service/legacysync/epoch_syncing.go:53]: Last successfully inserted block.

File: core/blockchain.go (hunk 13 above)
Comment (Frozen) [reply core/blockchain.go:1948]: Technically yes, but it would be inconvenient.

File: core/blockchain.go (hunk 12 above)
Comment (Frozen) [reply core/blockchain.go:366]: Added comment. 
	// Skip state check on initialization.
	// Should be enabled for epoch chain because it doesn't contain valid state.

File: api/service/legacysync/epoch_syncing.go (hunk 10 above)
Comment (Frozen) [reply api/service/legacysync/epoch_syncing.go]: fixed

File: node/node.go [hunk 14]
```diff
@@ -151,7 +152,11 @@ func (node *Node) Blockchain() *core.BlockChain {
 
 // Beaconchain returns the beaconchain from node.
 func (node *Node) Beaconchain() *core.BlockChain {
-	bc, err := node.shardChains.ShardChain(shard.BeaconChainShardID)
+	return node.EpochChain()
```
Comment (rlan35) [inline node/node.go:155]: Please don't change the existing Beaconchain() which is used somewhere else which may require the full chain. Beaconchain should mean the full chain while EpochChain means the chain with only epoch blocks. Beaconchain can be used as a EpochChain, but EpochChain can't be used as Beaconchain.

File: node/node_handler.go [hunk 15]
```diff
@@ -66,25 +66,7 @@ func (node *Node) HandleNodeMessage(
 	case proto_node.Block:
 		switch blockMsgType := proto_node.BlockMessageType(msgPayload[0]); blockMsgType {
 		case proto_node.Sync:
-			blocks := []*types.Block{}
```
Comment (rlan35) [inline node/node_handler.go:69]: We may still need some logic in this part because instead of broadcasting every block from beacon chain, now only the epoch blocks need to be broadcasted and processed here.

File: node/node.go (hunk 14 above)
Comment (Frozen) [reply node/node.go]: Fixed

**Comment by sophoah** [comment]
@Frozen in regards of the recent S1/S2 consensus loss, can you explain how the s0 last epoch sync process starts ? is it possible for a non S0 validator node currently up to date, later lagging behind to the last s0 block, to start a sync when new epoch starts ?

This will help to minimize the time for a shard to recover the consensus when the node is lagging behind at epoch change

cc @rlan35

**Comment by sophoah** [comment]
With the latest commit, as of now :
- node will sync the last epoch block of the previous epoch
- node will be stuck at that block and won't further sync
- and it happens at each new epoch 

is that an intended behavior ? or shouldn't we sync all the block in the current epoch including the last block of the previous ?

**Comment by sophoah** [comment]
question : would the code work as well with streamsync enabled ?

**Comment by Frozen** [comment]
No, it doesn't support streamsync.

**Comment by Frozen** [comment]
Right now it downloads all epoch blocks only, mean last blocks of epoch, with shard state info. i'm going to change it to download only last block of previous epoch, it will reduce lag between node node is ready and not to validate new blocks.

**Comment by rlan35** [comment]
> Right now it downloads all epoch blocks only, mean last blocks of epoch, with shard state info. i'm going to change it to download only last block of previous epoch, it will reduce lag between node node is ready and not to validate new blocks.

All the epoch blocks starting from genesis should be downloaded because you need to be able to verify each of them. Without the previous epoch block, you don't have the committee information to verify the block successfully. So I think you need to make sure the verification of a new epoch block is based on committee of last epoch block.

**Comment by sophoah** [comment]
Hit an issue with the last epoch change, not sure if that happened before as well but the test node is currently over 500 blocks after the new epoch blocks (since epoch change) and the node hasn't downloaded the new epoch block.

here is a log that might be a trigger of this issue :
```json
{"level":"error","error":"empty payload","caller":"/home/sp/harmony/harmony/api/service/legacysync/epoch_syncing.go:145","time":"2022-04-10T23:57:17.257374782-05:00","message":"[SYNC] ProcessStateSync failed (isBeacon: true, ShardID: 0, otherHeight: 18446744073709551615, currentHeight: 25100287)"}
{"level":"info","caller":"/home/sp/harmony/harmony/api/service/legacysync/epoch_syncing.go:121","time":"2022-04-10T23:57:19.260001879-05:00","message":"[SYNC] Node is OUT OF SYNC (isBeacon: true, ShardID: 0, otherHeight: 18446744073709551615, currentHeight: 25100287)"}
{"level":"error","error":"empty payload","caller":"/home/sp/harmony/harmony/api/service/legacysync/epoch_syncing.go:145","time":"2022-04-10T23:57:19.260091004-05:00","message":"[SYNC] ProcessStateSync failed (isBeacon: true, ShardID: 0, otherHeight: 18446744073709551615, currentHeight: 25100287)"}
{"level":"info","caller":"/home/sp/harmony/harmony/api/service/legacysync/epoch_syncing.go:121","time":"2022-04-10T23:57:21.26087363-05:00","message":"[SYNC] Node is OUT OF SYNC (isBeacon: true, ShardID: 0, otherHeight: 18446744073709551615, currentHeight: 25100287)"}
{"level":"error","error":"empty payload","caller":"/home/sp/harmony/harmony/api/service/legacysync/epoch_syncing.go:145","time":"2022-04-10T23:57:21.260957516-05:00","message":"[SYNC] ProcessStateSync failed (isBeacon: true, ShardID: 0, otherHeight: 18446744073709551615, currentHeight: 25100287)"}
{"level":"info","caller":"/home/sp/harmony/harmony/api/service/legacysync/epoch_syncing.go:121","time":"2022-04-10T23:57:23.261946835-05:00","message":"[SYNC] Node is OUT OF SYNC (isBeacon: true, ShardID: 0, otherHeight: 18446744073709551615, currentHeight: 25100287)"}
```

**Comment by sophoah** [comment]
@PkayJava could you help to test that PR on a testnet S0 node and observe the behavior ?
- build a testnet s0 node and validator (I can send you testnet token to create the validator)
- start it with an empty shard 0 DB
- observe sync and consensus
report here any issue. Thanks

**Comment by PkayJava** [comment]
@sophoah - noted with okay.

**Comment by PkayJava** [comment]
@Konstantin - This PR look good.

{"level":"info","blockNum":24600427,"epochNum":75664,"ViewId":24606416,"blockHash":"0xc1a4f16aec3131347ee4424213986f3e85fdcfe23eb2d20b6470b3b6f5e67769","numTxns":1,"numStakingTxns":0,"numSignatures":1,"caller":"/opt/go-path/src/github.com/harmony-one/harmony/node/node_handler.go:377","time":"2022-04-28T15:07:25.345778224Z","message":"BINGO !!! Reached Consensus"}

File: api/service/legacysync/epoch_syncing.go [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+func (ss *EpochSync) isInSync(doubleCheck bool) SyncCheckResult {
+	if ss.syncConfig == nil {
+		return SyncCheckResult{} // If syncConfig is not instantiated, return not in sync
+	}
+	otherHeight1 := getMaxPeerHeight(ss.syncConfig)
+	curBlock := ss.beaconChain.CurrentBlock()
+	epochLastBlock := shard.Schedule.EpochLastBlock(curBlock.Epoch().Uint64())
+	lastHeight := ss.beaconChain.CurrentBlock().NumberU64()
+
+	if !doubleCheck {
```
Comment (rlan35) [inline api/service/legacysync/epoch_syncing.go:56]: Whether the epoch chain is in sync should be determined by this:

1. Get my latest epoch block (note since it's epoch chain, the CurrentBlock() will return an epoch block (the last block of an epoch). The epoch of this block is my_epoch.
2. Get the highest block from Peer. The epoch of this block is their_epoch.
3. Compare whether my_epoch == their_epoch-1.  Note it's almost impossible to have my_epoch == their_epoch because I only have the epoch block which is the last block of an epoch. So when I sync'ed the epoch block of an epoch, the beacon chain should already moved to the next epoch.

File: api/service/legacysync/epoch_syncing.go [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+			Uint64("OtherHeight", otherHeight1).
+			Uint64("lastHeight", lastHeight).
+			Msg("[EPOCHSYNC] Checking sync status")
+		return SyncCheckResult{
+			IsInSync:    inSync,
+			OtherHeight: otherHeight1,
+			HeightDiff:  heightDiff,
+		}
+	}
+	// double check the sync status after 1 second to confirm (avoid false alarm)
```
Comment (rlan35) [inline api/service/legacysync/epoch_syncing.go:72]: No need for double check since epoch block only increase every 18.5hours. There is almost no chance of false alarm.

File: api/service/legacysync/epoch_syncing.go [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+		timeout = ss.syncLoop(bc, worker, isBeacon, consensus)
+	}
+}
+
+func (ss *EpochSync) syncLoop(bc *core.BlockChain, worker *worker.Worker, isBeacon bool, _ *consensus.Consensus) (timeout int) {
+	maxHeight := getMaxPeerHeight(ss.syncConfig)
+	for {
+		block := bc.CurrentBlock()
+		height := block.NumberU64()
+		if height >= maxHeight {
```
Comment (rlan35) [inline api/service/legacysync/epoch_syncing.go:113]: again, this sync check doesn't work. The epoch block will always be less than the maxHeight since it's the last block of an epoch. We should check my_epoch >= theirEpoch - 1.

File: api/service/legacysync/epoch_syncing.go [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+
+		var heights []uint64
+		curEpoch := block.Epoch()
+		for i := 0; i < int(SyncLoopBatchSize); i++ {
+			epochLastBlock := shard.Schedule.EpochLastBlock(curEpoch.Uint64())
+			if epochLastBlock > maxHeight {
+				break
+			}
+			if height != epochLastBlock {
+				heights = append(heights, epochLastBlock)
```
Comment (rlan35) [inline api/service/legacysync/epoch_syncing.go:132]: when i = 0, we are downloading the latest epoch block we already had "block := bc.CurrentBlock()". This is an epoch block and if we get EpochLastBlock of it's epoch. it returns the block itself. So we should skip this first one.

File: api/service/legacysync/helpers.go [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+
+	"github.com/ethereum/go-ethereum/common/math"
+	"github.com/harmony-one/harmony/api/service/legacysync/downloader"
+	"github.com/harmony-one/harmony/internal/utils"
+	"github.com/harmony-one/harmony/p2p"
+	"github.com/pkg/errors"
+)
+
+// getMaxPeerHeight gets the maximum blockchain heights from peers
+func getMaxPeerHeight(syncConfig *SyncConfig) uint64 {
```
Comment (rlan35) [inline api/service/legacysync/helpers.go:15]: Just to confirm. the code in this file helpers.go is directly copy from the original code without functional change, right?

File: api/service/legacysync/epoch_syncing.go (hunk 16 above)
Comment (sophoah) [reply api/service/legacysync/epoch_syncing.go:56]: just a side node on this comment @rlan35, it happened on the mainnet i am testing that we have this logs constantly :
```
{"level":"info","caller":"/home/sp/harmony/harmony/api/service/legacysync/epoch_syncing.go:121","time":"2022-06-19T18:56:04.933824999-05:00","message":"[EPOCHSYNC] Node is OUT OF SYNC (isBeacon: true, ShardID: 0, otherHeight: 18446744073709551615, currentHeight: 27951103)"}
```
until a restart of the node is initiated, and the otherHeight is correct

Comment (Frozen) [reply api/service/legacysync/epoch_syncing.go:56]: 18446744073709551615 is max value for uint64

It started with this value, other peers should overwrite it, but it doesn't. Seems like no other peers or no even one peer responses correctly.

File: api/service/legacysync/helpers.go (hunk 20 above)
Comment (Frozen) [reply api/service/legacysync/helpers.go:15]: yes, copy paste

File: api/service/legacysync/epoch_syncing.go (hunk 17 above)
Comment (Frozen) [reply api/service/legacysync/epoch_syncing.go:72]: removed

File: api/service/legacysync/epoch_syncing.go [hunk 21]
```diff
@@ -0,0 +1,231 @@
...
+			Int("Peers count", ss.syncConfig.PeersCount()).
+			Msg("[EPOCHSYNC] No peers for get height")
+		return SyncCheckResult{}
+	}
+	curEpoch := ss.beaconChain.CurrentBlock().Epoch().Uint64()
+	otherEpoch := shard.Schedule.CalcEpochNumber(otherHeight1).Uint64()
+	normalizedOtherEpoch := otherEpoch - 1
+	inSync := curEpoch == normalizedOtherEpoch
+
+	heightDiff := normalizedOtherEpoch - curEpoch
```
Comment (rlan35) [inline api/service/legacysync/epoch_syncing.go:66]: This is not height diff anymore. it's epoch diff. Please change the naming.

File: api/service/legacysync/epoch_syncing.go [hunk 22]
```diff
@@ -0,0 +1,231 @@
...
+	}
+
+	utils.Logger().Info().
+		Uint64("OtherEpoch", otherEpoch).
+		Uint64("CurrentEpoch", curEpoch).
+		Msg("[EPOCHSYNC] Checking sync status")
+	return SyncCheckResult{
+		IsInSync:    inSync,
+		OtherHeight: otherHeight1,
+		HeightDiff:  heightDiff,
```
Comment (rlan35) [inline api/service/legacysync/epoch_syncing.go:78]: EpochDiff.

File: node/node_syncing.go [hunk 23]
```diff
@@ -224,32 +220,18 @@ func (node *Node) doBeaconSyncing() {
...
-						}
-						err := node.beaconSync.UpdateBlockAndStatus(
-							beaconBlock, node.Beaconchain(), true,
-						)
-						if err != nil {
-							node.beaconSync.AddLastMileBlock(beaconBlock)
-						}
-					}
-				}
+			for _ = range node.BeaconBlockChannel {
```
Comment (rlan35) [inline node/node_syncing.go:223]: Why is this loop empty now?

File: api/service/legacysync/epoch_syncing.go [hunk 24]
```diff
@@ -0,0 +1,231 @@
...
+		utils.Logger().Info().
+			Msgf("[EPOCHSYNC] Node is OUT OF SYNC (isBeacon: %t, ShardID: %d, otherEpoch: %d, currentEpoch: %d, peers count %d)",
+				isBeacon, bc.ShardID(), otherEpoch, curEpoch, ss.syncConfig.PeersCount())
+
+		var heights []uint64
+		loopEpoch := curEpoch + 1
+		for len(heights) < int(SyncLoopBatchSize) {
+			epochLastBlockHeight := shard.Schedule.EpochLastBlock(loopEpoch)
+			if epochLastBlockHeight > maxHeight {
+				break
```
Comment (rlan35) [inline api/service/legacysync/epoch_syncing.go:123]: Change the condition to check on epoch: loopEpoch >= otherEpoch

File: internal/utils/crosslink_storage.go [hunk 25]
```diff
@@ -0,0 +1,27 @@
+package utils
+
+import "sync"
+
+type CrosslinkStorage struct {
```
Comment (rlan35) [inline internal/utils/crosslink_storage.go:5]: What's CrosslinkStorage for in this PR? Please delete if it's out of date.

File: node/node_cross_link.go [hunk 26]
```diff
@@ -58,7 +62,82 @@ func (node *Node) ProcessCrossLinkHeartbeatMessage(msgPayload []byte) {
...
+		}
+	}
+
+	if !keyExists {
+		return errors.New("pub key doesn't exist")
+	}
+
+	utils.Logger().Info().
+		Msgf("[ProcessCrossLinkHeartbeatMessage] storing latestSentCrosslink to %d", hb.LatestContinuousBlockNum)
+	atomic.StoreUint64(&latestSentCrosslink, hb.LatestContinuousBlockNum)
```
Comment (rlan35) [inline node/node_cross_link.go:139]: latestSentCrosslink is just a temporary in-memory cache. The real lastContinuousCrossLink should be stored in DB.  Take a look at "ReadShardLastCrossLink" in the original code.

File: node/node_handler.go [hunk 27]
```diff
@@ -84,9 +84,11 @@ func (node *Node) HandleNodeMessage(
 						if block.ShardID() == 0 {
 							utils.Logger().Info().
 								Msgf("Beacon block being handled by block channel: %d", block.NumberU64())
-							go func(blk *types.Block) {
-								node.BeaconBlockChannel <- blk
-							}(block)
+							if block.IsLastBlockInEpoch() {
+								go func(blk *types.Block) {
+									node.BeaconBlockChannel <- blk
```
Comment (rlan35) [inline node/node_handler.go:89]: Note to also change the beacon chain block broadcasting logic to only broadcast epoch block.

File: node/node_handler.go [hunk 28]
```diff
@@ -177,6 +179,9 @@ func (node *Node) BroadcastCrossLinkFromShardsToBeacon() { // leader of 1-3 shar
 	if node.IsRunningBeaconChain() {
 		return
 	}
+	if !(node.Consensus.IsLeader() || rand.Intn(100) == 0) {
```
Comment (rlan35) [inline node/node_handler.go:182]: Change to do 2%. "rand.Intn(100) <= 1"

Comment (Frozen) [reply node/node_handler.go:182]: fixed

File: internal/utils/crosslink_storage.go (hunk 25 above)
Comment (Frozen) [reply internal/utils/crosslink_storage.go:5]: removed

File: api/service/legacysync/epoch_syncing.go (hunk 21 above)
Comment (Frozen) [reply api/service/legacysync/epoch_syncing.go:66]: fixed

File: api/service/legacysync/epoch_syncing.go (hunk 22 above)
Comment (Frozen) [reply api/service/legacysync/epoch_syncing.go:78]: SyncCheckResult is struct with field HeightDiff, and it used in other places too, for example in legasy sync, where 
HeightDiff is appropriate name. I can't just rename it, it will lead to other changes. Suggest leave it as is.

File: node/node_syncing.go (hunk 23 above)
Comment (Frozen) [reply node/node_syncing.go:223]: everything inside this loop is moved to another place or removed as useless. But we still need to read from channel, or it will block. I will consider removing this channel, but not necessarily in the current pr.

File: node/node_cross_link.go (hunk 26 above)
Comment (rlan35) [reply node/node_cross_link.go:139]: If you don't store the latestContinuousBlockNum in DB, you should update the logic involving ReadShardLastCrossLink() since that part of the code is still reading that data from DB.

**Review by rlan35 — APPROVED** [review]

**Comment by rlan35** [comment]
Looks good. Please fix travis, and then let's test on testnet first.

File: api/service/legacysync/epoch_syncing.go [hunk 29]
```diff
@@ -0,0 +1,196 @@
...
+		if maxHeight == 0 || maxHeight == math.MaxUint64 {
+			utils.Logger().Info().
+				Msgf("[EPOCHSYNC] No peers to sync (isBeacon: %t, ShardID: %d, peerscount: %d)",
+					isBeacon, bc.ShardID(), ss.syncConfig.PeersCount())
+			return 10
+		}
+
+		curEpoch := bc.CurrentBlock().Epoch().Uint64()
+		otherEpoch := shard.Schedule.CalcEpochNumber(maxHeight).Uint64()
+		if otherEpoch-1 <= curEpoch {
```
[the remark this answers is not in the document]
Comment (rlan35) [reply api/service/legacysync/epoch_syncing.go:104]: Good point. better change it for peace of mind.

File: p2p/stream/protocols/sync/chain.go [hunk 30]
```diff
@@ -50,23 +50,20 @@ func (ch *chainHelperImpl) getBlockHashes(bns []uint64) []common.Hash {
 }
 
 func (ch *chainHelperImpl) getBlocksByNumber(bns []uint64) ([]*types.Block, error) {
-	var (
-		blocks = make([]*types.Block, 0, len(bns))
-	)
+	curBlock := ch.chain.CurrentHeader().Number().Uint64()
```
Comment (rlan35) [inline p2p/stream/protocols/sync/chain.go:53]: Is this change a bug fix?

**Review by rlan35 — APPROVED** [review]

**Comment by rlan35** [comment]
Finally looks good overall!

File: api/service/legacysync/epoch_syncing.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (Frozen) [reply api/service/legacysync/epoch_syncing.go]: fixed

**Review by MaxMustermann2 — APPROVED** [review]
